### ISSUE TITLE
How to use cryodrgn for homogeneous reconstruction with consensus poses?

### ISSUE THREAD

**Issue by Chen-Suyi** [ml-struct-bio/cryodrgn#445]
Hi CryoDRGN team,

Thank you for your great work. I know cryodrgn can perform ab initio homogeneous reconstruction. However, what I want to do is homogeneous reconstruction with given consensus poses. Is there a command to achieve this without modifying the source code?

**Comment by michal-g** [comment]
Hi, I think you are looking for the `cryodrgn train_nn` command — I've posted the docstring below, accessible through the `-h` flag, but let us know if you have any questions about how to use it!

```bash
$ cryodrgn train_nn -h
usage: cryodrgn train_nn [-h] -o OUTDIR --poses POSES [--ctf pkl] [--load WEIGHTS.PKL] [--checkpoint CHECKPOINT] [--log-interval LOG_INTERVAL] [-v] [--seed SEED] [--shuffle-seed SHUFFLE_SEED] [--uninvert-data] [--no-window] [--window-r WINDOW_R] [--ind IND] [--lazy] [--shuffler-size SHUFFLER_SIZE] [--datadir DATADIR] [-n NUM_EPOCHS] [-b BATCH_SIZE]
                         [--wd WD] [--lr LR] [--norm NORM NORM] [--no-amp] [--multigpu] [--do-pose-sgd] [--pretrain PRETRAIN] [--emb-type {s2s2,quat}] [--pose-lr POSE_LR] [--layers LAYERS] [--dim DIM] [--l-extent L_EXTENT] [--pe-type {geom_ft,geom_full,geom_lowf,geom_nohighf,linear_lowf,gaussian,none}] [--pe-dim PE_DIM] [--domain {hartley,fourier}]
                         [--activation {relu,leaky_relu}] [--feat-sigma FEAT_SIGMA]
                         particles

Train a neural net to model a 3D density map given 2D images with pose assignments.

positional arguments:
  particles             Input particles (.mrcs, .star, .cs, or .txt)

options:
  -h, --help            show this help message and exit
  -o OUTDIR, --outdir OUTDIR
                        Output directory to save model
  --poses POSES         Image poses (.pkl)
  --ctf pkl             CTF parameters (.pkl)
  --load WEIGHTS.PKL    Initialize training from a checkpoint
  --checkpoint CHECKPOINT
                        Checkpointing interval in N_EPOCHS (default: 1)
  --log-interval LOG_INTERVAL
                        Logging interval in N_IMGS (default: 1000)
  -v, --verbose         Increase verbosity
  --seed SEED           Random seed
  --shuffle-seed SHUFFLE_SEED
                        Random seed for data shuffling

Dataset loading:
  --uninvert-data       Do not invert data sign
  --no-window           Turn off real space windowing of dataset
  --window-r WINDOW_R   Windowing radius (default: 0.85)
  --ind IND             Filter particle stack by these indices
  --lazy                Lazy loading if full dataset is too large to fit in memory
  --shuffler-size SHUFFLER_SIZE
                        If non-zero, will use a data shuffler for faster lazy data loading.
  --datadir DATADIR     Path prefix to particle stack if loading relative paths from a .star or .cs file

Training parameters:
  -n NUM_EPOCHS, --num-epochs NUM_EPOCHS
                        Number of training epochs (default: 20)
  -b BATCH_SIZE, --batch-size BATCH_SIZE
                        Minibatch size (default: 8)
  --wd WD               Weight decay in Adam optimizer (default: 0)
  --lr LR               Learning rate in Adam optimizer (default: 0.0001)
  --norm NORM NORM      Data normalization as shift, 1/scale (default: mean, std of dataset)
  --no-amp              Do not use mixed-precision training
  --multigpu            Parallelize training across all detected GPUs

Pose SGD:
  --do-pose-sgd         Refine poses with gradient descent
  --pretrain PRETRAIN   Number of epochs with fixed poses before pose SGD (default: 5)
  --emb-type {s2s2,quat}
                        SO(3) embedding type for pose SGD (default: quat)
  --pose-lr POSE_LR     Learning rate for pose optimizer (default: 0.0001)

Network Architecture:
  --layers LAYERS       Number of hidden layers (default: 3)
  --dim DIM             Number of nodes in hidden layers (default: 1024)
  --l-extent L_EXTENT   Coordinate lattice size (if not using positional encoding) (default: 0.5)
  --pe-type {geom_ft,geom_full,geom_lowf,geom_nohighf,linear_lowf,gaussian,none}
                        Type of positional encoding (default: gaussian)
  --pe-dim PE_DIM       Num frequencies in positional encoding (default: D/2)
  --domain {hartley,fourier}
                        Volume decoder representation (default: fourier)
  --activation {relu,leaky_relu}
                        Activation (default: relu)
  --feat-sigma FEAT_SIGMA
                        Scale for random Gaussian features (default: 0.5)

Example usage
-------------
$ cryodrgn train_nn projections.mrcs --poses angles.pkl --ctf ctf.pkl \
                                     -o output/train_nn -n 10

# Run with more epochs
$ cryodrgn train_nn projections.star --poses angles.pkl --ctf.pkl \
                                     -o outs/003_train-nn --num-epochs 30 --lr 0.01

# Restart after already running the same command with some epochs completed
$ cryodrgn train_nn projections.star --poses angles.pkl --ctf.pkl \
                                     -o outs/003_train-nn --num-epochs 75 --lr 0.01 \
                                     --load latest

```

**Comment by Chen-Suyi** [comment]
Thank you for your prompt reply. I have another question. Is it possible to build half maps and calculate fsc like backproject_voxel without modifying the source code? I did not see that option.

**Comment by zhonge** [comment]
Chiming in here briefly -- the reason that we don't support it is because half-map FSCs are not a valid approach for resolution estimation.

**Comment by Chen-Suyi** [comment]
> Chiming in here briefly -- the reason that we don't support it is because half-map FSCs are not a valid approach for resolution estimation.

Is there a better way to measure the resolution of consensus maps? I know half-map FSCs measure self-consistency, and typically depend on the accuracy of poses. However, to my best knowledge, there is no better way for resolution estimation.

**Comment by Chen-Suyi** [comment]
> Although we do not presently support generating half-maps and calculating FSCs directly as part of neural-net-based reconstruction commands like `train_vae` and `abinit_homo` (unlike `backproject_voxel` where as you noted this can be done within the command itself) you can instead use our collection of utility commands to produce half-maps and calculate FSCs between the outputs of the two separate reconstruction tasks:
> 
> # Create filtering indices corresponding to a bipartition of the image stack, assuming 100,000 images
> $ cryodrgn_utils select_random 100000 -o ind_halfmap_a.pkl -s ind_halfmap_b.pkl --frac 0.5
> 
> # Train fixed-pose reconstruction using both halves
> $ cryodrgn train_nn particles.128.mrcs --poses pose.pkl --ctf ctf.pkl --ind ind_halfmap_a.pkl -o outs/hmap.a/ ...
> $ cryodrgn train_nn particles.128.mrcs --poses pose.pkl --ctf ctf.pkl --ind ind_halfmap_b.pkl -o outs/hmap.b/ ...
> 
> # Calculate FSC between output volumes
> $ cryodrgn_utils fsc outs/hmap.a/reconstruct.mrc outs/hmap.b/reconstruct.mrc

Thank you for your guidance. I noticed that 50 epochs should be used for cryodrgn to build consensus maps from EMPIAR-10028 and EMPIAR-10049. Nevertheless, the default setting in the source code is 20 epochs. Is that 20 epochs enough for such a task? I followed the README, so the resolution of input particles is 256*256.

**Comment by Chen-Suyi** [comment]
Thank you for your suggestions. The output volume at the 20th epoch converged, but still had artifacts. I am not sure how many epoch is needed to reach the best performance. I will try as many epochs as possible. I do not think there is such a concept of overfitting for homogeneous reconstruction.
